### PR TITLE
feat(theming): Added support to generate theme files

### PR DESCRIPTION
**Pull request by gregroyal**

### CODE REVIEW

File: test/plugins/scss/index.test.js [hunk 1]
```diff
@@ -10,6 +10,7 @@ const rimraf = require('rimraf');
 describe('scss resolver', () => {
   afterEach(() => {
     scss.clear();
+    rimraf.sync(path.join(outputDir + '/themes', '*'));
```
Comment (wallw-bits) [inline test/plugins/scss/index.test.js]: The path call there should be `path.join(outputDir, 'themes', '*')`

File: plugins/scss/index.js [hunk 2]
```diff
@@ -154,14 +159,54 @@ const writer = function(thisPackage, outputDir) {
...
-      promises.push(concat(scssEntries, file));
+      promises.push(concat.concatAsync(scssEntries, file));
     }
 
     file = path.resolve(outputDir, 'node-sass-args');
     console.log('Writing ' + file);
     promises.push(fs.writeFileAsync(file, args.join(' ')));
 
-    return Promise.all(promises);
+    return new Promise(function(resolve, reject) {
```
Comment (wallw-bits) [inline plugins/scss/index.js]: Why the extra `new Promise()` wrapper here? Shouldn't you be able to simply return the `Promise.all(...)`?

File: plugins/scss/index.js [hunk 3]
```diff
@@ -154,14 +159,54 @@ const writer = function(thisPackage, outputDir) {
...
+        // Write a file for each theme also
+        if (utils.isAppPackage(thisPackage) &&
+            thisPackage.build.scss &&
+            thisPackage.build.themes &&
+            thisPackage.build.themes.length) {
+          // Take off the combined.scss generic file
+          args.pop();
+
+          var themeDir = outputDir + '/themes';
+          mkdirp.sync(themeDir);
```
Comment (wallw-bits) [inline plugins/scss/index.js]: It is considered poor practice in Node to use synchronous calls. Please consider using the async call instead. You may want to promisify it similarly to how `concat` was done above.

File: plugins/scss/index.js [hunk 4]
```diff
@@ -154,14 +159,54 @@ const writer = function(thisPackage, outputDir) {
...
+
+          thisPackage.build.themes.push('default');
+
+          // For all the themes, create a combined.scss and a node-sass-args file
+          thisPackage.build.themes.forEach(function(theme) {
+            var themeOutput = path.resolve(themeDir, theme + '.combined.scss');
+            var themeArgs = path.resolve(themeDir, theme + '.node-sass-args');
+            promises.push(fs.writeFileAsync(themeArgs, args.join(' ') + ' ' + themeOutput));
+            promises.push(new Promise(function(resolve, reject) {
```
Comment (wallw-bits) [inline plugins/scss/index.js]: Same here. There is no need to make a `new Promise()` wrapper when you could return the `fs.writeFileAsync()` promise directly.

File: plugins/scss/index.js [hunk 5]
```diff
@@ -154,14 +159,54 @@ const writer = function(thisPackage, outputDir) {
...
+      promises.push(concat.concatAsync(scssEntries, file));
     }
 
     file = path.resolve(outputDir, 'node-sass-args');
     console.log('Writing ' + file);
     promises.push(fs.writeFileAsync(file, args.join(' ')));
 
-    return Promise.all(promises);
+    return new Promise(function(resolve, reject) {
+      resolve(Promise.all(promises).then(function() {
```
Comment (wallw-bits) [inline plugins/scss/index.js]: I know we don't have a good lint rule for this, but please move `.then()` and other chained promise calls to their own line to match up with project conventions.

File: plugins/scss/index.js [hunk 6]
```diff
@@ -154,14 +159,54 @@ const writer = function(thisPackage, outputDir) {
...
+
+          thisPackage.build.themes.push('default');
+
+          // For all the themes, create a combined.scss and a node-sass-args file
+          thisPackage.build.themes.forEach(function(theme) {
+            var themeOutput = path.resolve(themeDir, theme + '.combined.scss');
+            var themeArgs = path.resolve(themeDir, theme + '.node-sass-args');
+            promises.push(fs.writeFileAsync(themeArgs, args.join(' ') + ' ' + themeOutput));
+            promises.push(new Promise(function(resolve, reject) {
+                resolve(fs.readFileAsync(path.resolve(outputDir, 'combined.scss'), 'utf8').then(function(fileContents) {
```
Comment (wallw-bits) [inline plugins/scss/index.js]: Same here with the chained promise call convention.

File: plugins/scss/index.js (hunk 3 above)
Comment (gregroyal) [reply plugins/scss/index.js]: I'm not sure if theres an async method? https://github.com/substack/node-mkdirp/blob/master/index.js

File: plugins/scss/index.js (hunk 2 above)
Comment (gregroyal) [reply plugins/scss/index.js]: I think i was trying to workaround the writeFileAsync still working when the test was trying to clean it up

File: plugins/scss/index.js (hunk 5 above)
Comment (gregroyal) [reply plugins/scss/index.js]: sure

File: plugins/scss/index.js (hunk 6 above)
Comment (gregroyal) [reply plugins/scss/index.js]: ok

File: test/plugins/scss/index.test.js (hunk 1 above)
Comment (gregroyal) [reply test/plugins/scss/index.test.js]: ok!

**Comment by gregroyal** [comment]
Addressed all your comments except the mkdirp one (which didnt seem like it had an async method) and works fine

**Comment by wallw-bits** [comment]
You might want to have a look at the [example from the docs](https://www.npmjs.com/package/mkdirp#example)

**Comment by gregroyal** [comment]
Should be good now!

**Pull request merged**